### PR TITLE
Bug 1204152 - Migrate FTU to l20n

### PR DESCRIPTION
**Pull request by zbraniecki**
https://bugzilla.mozilla.org/show_bug.cgi?id=1204152

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=607cfc9d5bf5461e3d9301797d515b584c9b0ec0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=683d22aa41c334fd48cd338a5f83622371ee8bb9

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c9a662e0db6e4636bd7189d322c3713e7f9910f0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cdde2583e74dd714b999ab105d480f190ed394fb

File: apps/ftu/js/app.js [hunk 1]
```diff
@@ -132,4 +127,12 @@ navigator.mozL10n.ready(function showBody() {
...
     }
   });
+}
+
+document.l10n.ready.then(() => {
+  window.performance.mark('l10nready');
+  init();
+
+  // we want to fire this function on each consequent language change
+  document.addEventListener('DOMRetranslated', init);
```
Comment (stasm) [inline apps/ftu/js/app.js]: You can safely register the listener outside of `l10n.ready`.

File: shared/js/tz_select.js [hunk 2]
```diff
@@ -122,23 +122,58 @@ function tzSelect(regionSelector, citySelector, onchange, onload) {
...
+      selector.innerHTML = '';
+
+      const collator = Intl.Collator(navigator.languages, {
+        usage: 'sort',
+      });
+
+      const frag = document.createDocumentFragment();
+      const itemArr = [];
+
+      options.forEach(option => {
```
Comment (stasm) [inline shared/js/tz_select.js]: How about:

    const items = options.map(…);

?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=aca954d8397d36c1f3f46a78577af1b07a5e9c21

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=53524c10a5e700616f8ef2a1775972daa2a11482

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ab4aa9a0a68b3f73c5d1bc3384f3decb8fa86798

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=04cde2c8636e79958e6fa5443d777ef630ffb2cf

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3e0fee985f47b9a03a8b92bad672e9c288f47128

File: apps/ftu/js/sim_manager.js [hunk 3]
```diff
@@ -140,14 +140,14 @@ var SimManager = (function() {
         UIManager.xckInput.value = '';
         UIManager.xckInput.classList.add('onerror');
         UIManager.xckError.classList.remove('hidden');
-        navigator.mozL10n.setAttributes(UIManager.xckRetriesLeft,
+        document.l10n.setAttributes(UIManager.xckRetriesLeft,
                                           'inputCodeRetriesLeft', l10nArgs);
```
Comment (stasm) [inline apps/ftu/js/sim_manager.js]: Nit: indentation.

File: apps/ftu/js/sim_manager.js [hunk 4]
```diff
@@ -118,14 +118,14 @@ var SimManager = (function() {
         UIManager.pukInput.classList.add('onerror');
         UIManager.pukError.classList.remove('hidden');
         UIManager.pukInfo.classList.add('hidden');
-        navigator.mozL10n.setAttributes(UIManager.pukRetriesLeft,
+        document.l10n.setAttributes(UIManager.pukRetriesLeft,
                                           'inputCodeRetriesLeft', l10nArgs);
```
Comment (stasm) [inline apps/ftu/js/sim_manager.js]: Nit: indentation.

File: apps/ftu/js/sim_manager.js (hunk 3 above)
Comment (stasm) [inline apps/ftu/js/sim_manager.js]: (And in other places in this file.)

File: apps/ftu/test/marionette/language_pseudo_localization_test.js [hunk 5]
```diff
@@ -45,8 +45,8 @@ marionette('First Time Use > Pseudo Localization', function() {
     ftu.selectLanguage('fr-x-psaccent');
 
     var translatedHeader = client.executeScript('' +
-      'var qps = window.wrappedJSObject.navigator.mozL10n.qps;' +
-      'return qps["fr-x-psaccent"].translate("Language");'
+      'var pseudo = window.wrappedJSObject.document.l10n.pseudo;' +
+      'return pseudo["fr-x-psaccent"].translate("Language");'
```
Comment (stasm) [inline apps/ftu/test/marionette/language_pseudo_localization_test.js]: `translate` is called `processString` now.

Comment (stasm) [inline apps/ftu/test/marionette/language_pseudo_localization_test.js]: And it's async.  Did this pass the tests?

File: shared/js/language_list.js [hunk 6]
```diff
@@ -52,32 +52,37 @@ exports.LanguageList = {
...
-        isCurrent = (lang === currentLang);
-        if (isCurrent || pseudoEnabled) {
-          languages[lang] = navigator.mozL10n.qps[lang].name;
-        }
-      }
+    if (!document.l10n) {
+      return Promise.resolve(languages);
     }
 
+    // 2. add the remaining runtime pseudolocales if qps enabled
```
Comment (stasm) [inline shared/js/language_list.js]: Remove the reference to `qps` in the comments in this file, please.

File: apps/ftu/js/app.js [hunk 7]
```diff
@@ -142,4 +137,11 @@ navigator.mozL10n.ready(function showBody() {
       }
     }
   });
+}
+
+document.l10n.ready.then(() => {
+  window.performance.mark('l10nready');
+  init();
```
Comment (stasm) [inline apps/ftu/js/app.js]: Do we need to wait for `document.l10n.ready` to call `init`?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dbea2df5c7506f40d073e861a2c9d587113e9ff4

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=41ac77b8551a55bfe682318400e1b6a09c4b9456

File: apps/ftu/test/marionette/lib/ftu.js [hunk 8]
```diff
@@ -31,12 +31,11 @@ Ftu.Selectors = {
 
 Ftu.prototype = {
   waitForL10nReady: function() {
-    this.client.helper.waitFor(function() {
-      return this.client.executeScript(function() {
-        return window.wrappedJSObject.navigator.mozL10n
-               .readyState === 'complete';
+    this.client.helper.waitFor(done => {
+      this.client.executeScript(function() {
```
Comment (stasm) [inline apps/ftu/test/marionette/lib/ftu.js]: Shouldn't this use `executeAsyncScript`?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6619cd1ddd0635149f1c8d7696e55d13e0b5b0e9

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f82c7a0afce229791fee7f2893f17d05db7dd758

File: apps/ftu/test/marionette/lib/ftu.js (hunk 8 above)
Comment (zbraniecki) [inline apps/ftu/test/marionette/lib/ftu.js]: I was able to get rid of `waitForL10nReady` overall :)

File: apps/ftu/test/marionette/rtl_test.js [hunk 9]
```diff
@@ -25,17 +25,23 @@ marionette('First Time Use >', function() {
...
                     .scriptWith(function(el) {
       return window.wrappedJSObject.getComputedStyle(el).direction;
     });
     assert.equal(direction, 'rtl');
-    assert.equal(translatedHeader, header.text());
+    quickly.executeAsyncScript(function() {
+      var pseudo = window.wrappedJSObject.document.l10n.pseudo;
+      pseudo['ar-x-psbidi'].processString('Language').then(
+        function(string) {
+          marionetteScriptFinished(string);
```
Comment (stasm) [inline apps/ftu/test/marionette/rtl_test.js]: Any specific reason not to use the fat arrow here?

    string => marionetteScriptFinished(string)

File: apps/ftu/test/unit/timezone_test.js [hunk 10]
```diff
@@ -79,7 +79,7 @@ suite('timezones >', function() {
     // Careful with the text transformations for the lists
     // we add 'tzRegion-' to mapped regions
     var mRegion = regionSelector.querySelectorAll('option')
-      [regionSelector.selectedIndex].textContent.split('-')[1];
+      [regionSelector.selectedIndex].getAttribute('data-l10n-id');
```
Comment (stasm) [inline apps/ftu/test/unit/timezone_test.js]: :+1:

File: apps/sharedtest/test/unit/language_list_test.js [hunk 11]
```diff
@@ -78,122 +77,160 @@ suite('Language List', function() {
...
 
   suite('extend language list with pseudolanguages', function() {
-    test('regular current language, pseudolocalization disabled', function() {
-      var langs = {
-        en: 'English'
+    setup(function() {
+      document.l10n = {
+        pseudo: {
+          'fr-x-psaccent': {
+            getName: function() {return 'Pseudo';}
```
Comment (stasm) [inline apps/sharedtest/test/unit/language_list_test.js]: Should this return `Promise.resolve('Pseudo')`?

Comment (zbraniecki) [inline apps/sharedtest/test/unit/language_list_test.js]: totally!

File: apps/ftu/test/marionette/rtl_test.js (hunk 9 above)
Comment (zbraniecki) [inline apps/ftu/test/marionette/rtl_test.js]: marionette can't handle fat arrow functions and it requires node 0.12 :(

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2258167d141000bb34ff9c52e5bdec737cbb056b

File: apps/ftu/test/marionette/rtl_test.js (hunk 9 above)
Comment (stasm) [inline apps/ftu/test/marionette/rtl_test.js]: That's indeed very specific. Thank you.

:)

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2258167d141000bb34ff9c52e5bdec737cbb056b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fa056de2dfffe7942f08f262f3fabace364a3fca

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=27cd0a17b2c13c43c506e843a1aa2aa91fae4b33

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=27cd0a17b2c13c43c506e843a1aa2aa91fae4b33

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a5ee8789bc08e74cee82f7109a63bc6baa9c8d67

**Pull request merged**